asapdiscovery/simulation/scripts/MproP0009.py

```diff
@@ -39,18 +32,67 @@ def get_args():
         default = "/data/chodera/lemonsk/asap-datasets/prepped_mpro_P0009/",
         help="Output simulation directory."
     )
+    parser.add_argument(
+        "-l",
+        "--ligand",
+        type = str,
+        default = "/data/chodera/lemonsk/covid-moonshot-ml/asapdiscovery/simulation/tests/inputs/MAT-POS-f2460aef-1.sdf",
+        help = "Ligand"
+    )
+    parser.add_argument(
+        "-p",
+        "--protein",
+        type = str,
+        default = "/data/chodera/lemonsk/asap-datasets/openmm_setup_processed/prepped_receptor_0-processed.pdb",
+        help = "Protein-receptor"
+    )
     args = parser.parse_args()
     return args
 
 args = get_args()
 
+from rdkit import Chem
+rdkitmol = Chem.SDMolSupplier(args.ligand)[0]
+rdkitmolh = Chem.AddHs(rdkitmol, addCoords=True);
+# ensure the chiral centers are all defined
+Chem.AssignAtomChiralTagsFromStructure(rdkitmolh);
+from openff.toolkit.topology import Molecule
+ligand_mol = Molecule(rdkitmolh);
+
+# Initialize a SystemGenerator
+from openmm import unit, app
+from openmmforcefields.generators import SystemGenerator
+forcefield_kwargs = {'constraints': app.HBonds, 'rigidWater': True, 'removeCMMotion': False, 'hydrogenMass': 4*unit.amu }
+periodic_forcefield_kwargs = {'nonbondedMethod': app.PME}
+system_generator = SystemGenerator(
+    forcefields=['amber/ff14SB.xml', 'amber/tip3p_standard.xml'],
+    small_molecule_forcefield='openff-1.3.1',
+    molecules=[ligand_mol], cache='cache.json',
+    forcefield_kwargs=forcefield_kwargs, periodic_forcefield_kwargs=periodic_forcefield_kwargs);
+
+# Use Modeller to combine the protein and ligand into a complex
+from openmm.app import PDBFile
+from openmm.app import Modeller
+protein_pdb = PDBFile(args.protein)
+modeller = Modeller(protein_pdb.topology, protein_pdb.positions)
+# This next bit is black magic.
+# Modeller needs topology and positions. Lots of trial and error found that this is what works to get these from
+# an openforcefield Molecule object that was created from a RDKit molecule.
+# The topology part is described in the openforcefield API but the positions part grabs the first (and only)
+# conformer and passes it to Modeller. It works. Don't ask why!
+modeller.add(ligand_mol.to_topology().to_openmm(), ligand_mol.conformers[0].to_openmm())
+
 # Input Files
 
-pdb = PDBFile(args.input_pdb_path)
+from openmm.app import ForceField
+#pdb = PDBFile(args.input_pdb_path)
 forcefield = ForceField('amber14-all.xml', 'amber14/tip3pfb.xml')
 
 # System Configuration
 
+from openmm.app import *
+from openmm.unit import * 
```

Comment:
I think the error is here. Your error message was
```
Traceback (most recent call last):
 File "/data/chodera/lemonsk/covid-moonshot-ml/asapdiscovery/simulation/scripts/MproP0009_2.py", line 137, in <module>
  modeller.addSolvent(forcefield, padding=0.9*nanometers, model='tip3p')
AttributeError: module 'openmm.app.modeller' has no attribute 'addSolvent'
```
And `modeller` shouldn't be a module, it should be an object.
But since you've used a star import after creating the modeller object, which overwrites the variable name (i.e. modeller is now a module name instead of an object name in the script namespace) later when you go to addSolvent, instead of using the object that you've already created, you're trying to call `openmm.app.modeller.addSolvent`, which doesn't exist. The function you'd like to use is `openmm.app.modeller.Modeller.addSolvent` which does exist.

http://docs.openmm.org/latest/api-python/generated/openmm.app.modeller.Modeller.html?highlight=openmm%20app%20modeller

https://www.geeksforgeeks.org/why-import-star-in-python-is-a-bad-idea/

Reply:
See, I know where you're coming from. I just haven't seen that before. Even Jenke had what I'm trying to do, but I'll try it and see. I just worry we have too many "modeller" objects out there, which could cause an issue? Maybe I'm being irrational here. But yes, no more star imports

Reply:
Right, this is Jenke's code:
```
from openmm.app import Modeller
modeller = Modeller(protein_pdb.topology, protein_pdb.positions)
# This next bit is black magic.
# Modeller needs topology and positions. Lots of trial and error found that this is what works to get these from
# an openforcefield Molecule object that was created from a RDKit molecule.
# The topology part is described in the openforcefield API but the positions part grabs the first (and only)
# conformer and passes it to Modeller. It works. Don't ask why!
modeller.add(ligand_mol.to_topology().to_openmm(), ligand_mol.conformers[0].to_openmm())

# We need to temporarily create a Context in order to identify molecules for adding virtual bonds
log.info(f':microscope:  Identifying molecules')
import openmm
integrator = openmm.VerletIntegrator(1*unit.femtoseconds)
system = system_generator.create_system(modeller.topology, molecules=ligand_mol)
context = openmm.Context(system, integrator, openmm.Platform.getPlatformByName('Reference'))
molecules_atom_indices = context.getMolecules()
del context, integrator, system

# Solvate
log.info(':droplet:  Adding solvent...')
# we use the 'padding' option to define the periodic box. The PDB file does not contain any
# unit cell information so we just create a box that has a 9A padding around the complex.
modeller.addSolvent(system_generator.forcefield, model='tip3p', padding=9.0*unit.angstroms)
```

In between creating the modeller object and adding solvent there are no star imports, i.e. there's no time for the name of the object to get overwritten!

Reply:
The problem isn't that star imports are always bad and wrong, its just that in this specific case your object is getting overwritten which is why you see the error message that you do!

Reply:
Simple example.
The first time, it at least knows what the function is, where as the second time it doesn't!
```
modeller = Modeller(pdb.topology, pdb.positions)
modeller.addSolvent()
Traceback (most recent call last):
  File "/Users/alexpayne/opt/anaconda3/envs/asap-simulation/lib/python3.11/site-packages/IPython/core/interactiveshell.py", line 3442, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-3b4dda05aca7>", line 1, in <module>
    modeller.addSolvent()
TypeError: Modeller.addSolvent() missing 1 required positional argument: 'forcefield'
from openmm.app import *
modeller.addSolvent()
Traceback (most recent call last):
  File "/Users/alexpayne/opt/anaconda3/envs/asap-simulation/lib/python3.11/site-packages/IPython/core/interactiveshell.py", line 3442, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-11-3b4dda05aca7>", line 1, in <module>
    modeller.addSolvent()
    ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'openmm.app.modeller' has no attribute 'addSolvent'
```